modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/ContextualTransactionValidator.scala

```diff
@@ -0,0 +1,282 @@
+          ParentOrdinalLowerThenLastProcessedTxOrdinal(signedTx.parent.ordinal, lastProcessedTransactionRef.ordinal).invalidNec
+        else if (signedTx.parent === lastProcessedTransactionRef) {
+          signedTx.validNec
+        } else {
+          val hasValidParent = txs.exists(_.exists {
+            case (parentOrdinal, parentTx) => parentTx.ref === signedTx.parent && parentOrdinal > lastProcessedTransactionRef.ordinal
+          })
+
+          if (hasValidParent) signedTx.validNec else HasNoMatchingParent(signedTx.parent.hash).invalidNec
+        }
+      }
+
+      private def validateNotLocked(
+        signedTx: Signed[Transaction]
+      ): ContextualTransactionValidationErrorOr[Signed[Transaction]] =
+        Validated
+          .condNec[ContextualTransactionValidationError, Signed[Transaction]](
+            !lockedAddresses.contains(signedTx.source),
+            signedTx,
+            LockedAddressError(signedTx.source)
+          )
+
+      private def recursivelyValidateCap(
+        minFee: TransactionFee,
+        ordinalDistanceForBaseBalance: Long,
+        baseBalance: Balance,
+        memPool: List[Hashed[Transaction]],
+        incomingTx: Hashed[Transaction],
+        balance: Balance,
+        lastMajorityTransactionOrdinal: SnapshotOrdinal,
+        currentOrdinal: SnapshotOrdinal
+      )(implicit next: Next[SnapshotOrdinal]): Option[Hashed[Transaction]] = {
+        @tailrec
+        def loop(
+          memPool: List[Hashed[Transaction]],
+          balance: Balance,
+          lastTxRefOrdinal: SnapshotOrdinal,
+          currentOrdinal: SnapshotOrdinal,
+          totalConsumedCap: Double,
+          toProcess: List[Hashed[Transaction]]
+        ): List[Hashed[Transaction]] = {
+          val distancePassed = currentOrdinal.value - lastTxRefOrdinal.value
+          val transactionsAllowedWithinInterval = balance.value.toDouble / baseBalance.value.toDouble
+          val distanceBetweenTxsNeeded = ordinalDistanceForBaseBalance.toDouble / transactionsAllowedWithinInterval
+          val consumedCap = distanceBetweenTxsNeeded / Math.max(distancePassed, 1)
+          val updatedTotalConsumedCap =
+            totalConsumedCap + (if (memPool.headOption.map(_.fee.value.toLong).getOrElse(0L) >= minFee.value) 0 else consumedCap)
+          val updatedBalance =
+            balance.minus(memPool.headOption.flatMap(tx => toAmount(tx.amount).plus(tx.fee).toOption).getOrElse(Balance.empty))
+
+          updatedBalance match {
+            case Left(_) =>
+              toProcess
+            case Right(_) if updatedTotalConsumedCap > 1 || memPool.isEmpty =>
+              toProcess
+            case Right(value) =>
+              loop(
+                memPool.tail,
+                value,
+                next.next(currentOrdinal),
+                currentOrdinal,
+                updatedTotalConsumedCap,
+                toProcess ++ memPool.headOption.toList
+              )
```

Comment:
- If we're concerned about throwing if list is empty (`memPool.headOption`) then aren't we concerned that `memPool.tail` will throw on empty? maybe empty `memPool` should be a loop-exit condition?
- If we're keeping this code, instead of constructing list from `headOption` maybe `memPool.headOption.fold(...)`  and using `:+`?
- Is `toProcess` going to be a long list? If so then maybe appending will be unnecessarily slow.

Reply:
>maybe empty memPool should be a loop-exit condition?

It is:
```
case Right(_) if updatedTotalConsumedCap > 1 || memPool.isEmpty =>
  toProcess
```
So I think we can safely use `memPool.head` there as `memPool.isEmpty` quits the loop.